pandas/core/algorithms.py

```diff
@@ -619,13 +619,19 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
 
     if sort and len(uniques) > 0:
         from pandas.core.sorting import safe_sort
-        try:
-            order = uniques.argsort()
-            order2 = order.argsort()
-            labels = take_1d(order2, labels, fill_value=na_sentinel)
-            uniques = uniques.take(order)
-        except TypeError:
-            # Mixed types, where uniques.argsort fails.
+        if na_sentinel == -1:
+            # GH-25409 take_1d only works for na_sentinels of -1
```

Comment:
Adding this if/else check to keep the usage of `take_1d` when possible. It was added some time ago to improve performance compared to `safe_sort`, but thus does not handle non-default na_sentinels. 
If we want to keep the code simpler, I can also remove the `take_1d` alltogether, but that would give a performance degradation compared to the last release.

Reply:
This seems reasonable. Alternatively, `safe_sort` could do this try / except?

Reply:
You mean the if / else? (as it would also have to deal with the non-default na_sentinels)

Reply:
Hmm yeah, it would need that as well. But I think it's probably fine here, unless we see this pattern coming up often when using safe_sort.